library/core/src/iter/traits/iterator.rs

```diff
@@ -1777,6 +1777,62 @@ pub trait Iterator {
         FromIterator::from_iter(self)
     }
 
+    /// Fallibly transforms an iterator into a collection, short circuiting if
+    /// a failure is encountered while collecting.
+    ///
+    /// `try_collect()` is essentially a wrapper around [`collect()`][`collect`] for types
+    /// that can already be fallibly constructed from iterators. Its main use case
+    /// is simplifying the conversion from iterators of [`Option<T>`][`Option`] into `Option<Vec<T>>`
+    /// (a similar technique exists for [`Result`]).
+    ///
+    /// In fact, `try_collect()` is generic over types that implement [`Try`], so any such type
+    /// can be fallibly collected into its output type in the same way.
+    ///
+    /// # Examples
+    /// Successfully collecting an iterator of `Option<i32>` into `Option<Vec<i32>>`:
+    /// ```
+    /// #![feature(iterator_try_collect)]
+    ///
+    /// let u = vec![Some(1), Some(2), Some(3)];
+    /// let v = u.into_iter().try_collect::<Vec<i32>>();
+    /// assert_eq!(v, Some(vec![1, 2, 3]));
+    /// ```
+    ///
+    /// Failing to collect in the same way:
+    /// ```
+    /// #![feature(iterator_try_collect)]
+    ///
+    /// let u = vec![Some(1), Some(2), None, Some(3)];
+    /// let v = u.into_iter().try_collect::<Vec<i32>>();
+    /// assert_eq!(v, None);
+    /// ```
+    ///
+    /// A similar example, but with `Result`:
+    /// ```
+    /// #![feature(iterator_try_collect)]
+    ///
+    /// let u: Vec<Result<i32, ()>> = vec![Ok(1), Ok(2), Ok(3)];
+    /// let v = u.into_iter().try_collect::<Vec<i32>>();
+    /// assert_eq!(v, Ok(vec![1, 2, 3]));
+    ///
+    /// let u = vec![Ok(1), Ok(2), Err(()), Ok(3)];
+    /// let v = u.into_iter().try_collect::<Vec<i32>>();
+    /// assert_eq!(v, Err(()));
+    /// ```
+    ///
+    /// [`collect`]: Iterator::collect
+    #[inline]
+    #[unstable(feature = "iterator_try_collect", issue = "none")]
```

Comment:
And yes, you'll open a "Library Tracking Issue" (it's one of the templates in https://github.com/rust-lang/rust/issues/new/choose ) at some point to put here.  But you don't need to do that until it's going to go in, so it's often reasonable to not bother doing so until you get a "sure, we can put something like this in nightly" from a reviewer.

Sometimes people are confident and open them along with the PR; sometimes the PRs are more speculative and they wait.  It's up to you.

Reply:
Alright, I'll probably hold off on opening anything until I've got a reasonable implementation then

Reply:
Oh, just `try_process(self, |i| i.collect())` worked?  That's convenient 🚀 

I think this looks good.  Can you open the tracking issue and update the attribute?  (And maybe squash the commits too, for cleanliness.)

Then once the check build is fixed -- I don't see how it could be this PR's fault -- I'll approve it.

Reply:
Sounds good! And yeah, I was impressed that it worked right off the bat too 😆